### PR TITLE
REF: share delete, putmask, insert between ndarray-backed EA indexes

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/arrays/_mixins.py [hunk 1]
```diff
@@ -44,6 +44,10 @@ def _box_func(self, x):
         """
         return x
 
+    def _validate_insert_value(self, value):
```
Comment (jreback) [inline pandas/core/arrays/_mixins.py:47]: should we just make this class an https://docs.python.org/3/library/abc.html then these aren't even needed

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -575,41 +564,30 @@ def shift(self, periods=1, freq=None):
...
-            if loc in (0, -len(self), -1, len(self) - 1):
-                freq = self.freq
-        else:
-            if is_list_like(loc):
-                loc = lib.maybe_indices_to_slice(
-                    np.asarray(loc, dtype=np.intp), len(self)
-                )
-            if isinstance(loc, slice) and loc.step in (1, None):
-                if loc.start in (0, None) or loc.stop in (len(self), None):
+        elif self.freq is not None:
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py:574]: i am not sure that these specific freq routines are really worth it given the extra code. Can't we just re-infer the freq? (sure if we do this constantly its not as performant, but this is not likey a common case).

ok for here, but if you can have a look at some point

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm ping on green (comments are for followon)

File: pandas/core/arrays/_mixins.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:47]: my hope is that we'll smooth out the small differences between the various `_validate_foo_value` methods (ATM trying to sort out Categorical's many special cases) and just have `_validate_setitemlike_value` or something like that

File: pandas/core/indexes/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py:574]: thats a reasonable thought.  i recently found a PITA bug in `infer_freq` though, so definitely not until that is sorted out

Comment (jreback) [reply pandas/core/indexes/datetimelike.py:574]: kk maybe worthile to create an issue to see what implications / address later is.

**Review by jreback — APPROVED** [review]

**Pull request merged**